src/compile.js

```diff
@@ -210,4 +210,9 @@ function compile(mediaType, apiElements, filename) {
   return { mediaType, transactions, annotations };
 }
 
+
+compile.compileBody = compileBody;
+compile.hasMultipartBody = hasMultipartBody;
```

Comment:
As I mentioned in sister PR [comment](https://github.com/apiaryio/dredd/pull/1094#discussion_r206821763), I think it would be good to emphasize that these are private/internal functions by leading underscores for example.

Reply:
Airbnb [prohibits](https://github.com/airbnb/javascript#naming--leading-underscore) underscores, so we'd need [ESLint exception](https://eslint.org/docs/rules/no-underscore-dangle) for that.

I decided to implement [the latest Airbnb JS filename conventions](https://github.com/airbnb/javascript#naming--filename-matches-export) in this new code, so people reading the code should only see something like:

```js
const getFooBar = require('./getFooBar');
getFooBar(42);
```

I hope such interface could help to make it clear the file does one thing and that one thing is required/imported. In the context of unit tests where the individual helpers do get required/imported, the intention could be still quite clear as the reader is... reading unit tests.

Where the intention isn't quite clear is the exports code itself. One solution could be a simple comment:

```js
}


// for the purposes of unit tests
compile.compileBody = compileBody;
compile.hasMultipartBody = hasMultipartBody;


module.exports = compile;
```

Underscores could be even more explicit, but require some additional work:

```js
}


// for the purpose of unit tests
compile._compileBody = compileBody;
compile._hasMultipartBody = hasMultipartBody;


module.exports = compile;
```

This also prevents property shadowing in case `compile` is an object or class, rather than a function. In this case ESLint needs to be silenced though:

```js
}


/* eslint-disable no-underscore-dangle */
compile._compileBody = compileBody;
compile._hasMultipartBody = hasMultipartBody;
/* eslint-enable no-underscore-dangle */


module.exports = compile;
```

The silly thing is I could not figure out how to add any extra information to the ESLint comment to let the reader know this is for the purposes of unit tests. If I add anything to the comments, they stop to get recognized by ESLint. The only solution would be to add an extra comment, which is really ugly:

```js
}


// for the purposes of unit tests
/* eslint-disable no-underscore-dangle */
compile._compileBody = compileBody;
compile._hasMultipartBody = hasMultipartBody;
/* eslint-enable no-underscore-dangle */


module.exports = compile;
```

Luckily, at least the imports/requires in unit tests do not need any special treatment regarding ESLint:

```js
const {
  _compileBody: compileBody,
  _hasMultipartBody: hasMultipartBody
} = require('../../src/compile');
```

A different solution for exports would be maybe to...

```js
// for the purposes of unit tests
compile._compileBody = compileBody; // eslint-disable-line no-underscore-dangle
compile._hasMultipartBody = hasMultipartBody; // eslint-disable-line no-underscore-dangle
```

Or maybe just this, since it's always going to be at the end of the file?

```js
/* exports for unit tests *//* eslint-disable */
compile._compileBody = compileBody;
compile._hasMultipartBody = hasMultipartBody;
```

Any thoughts on this? Do we want to do it this way? Is there an elegant solution to this?

Reply:
Personally, I like
```javascript
// For the purpose of unit tests
compile._compileBody = compileBody;
compile._hasMultipartBody = hasMultipartBody;
```
the most.

Regarding `no-underscore-dangle` eslint rule - do we need (want) it though? Just because Airbnb's JavaScript styleguide is considered the holy grail, we don't have to think about it as such (and because of decaff, we have different settings for some rules anyway).

On the other hand, we could (and probably should) think about _any_ change to _public_ contract as breaking - then something like
```javascript
// For the purpose of unit tests
compile.compileBody = compileBody;
compile.hasMultipartBody = hasMultipartBody;
```
would be perfectly fine.

Reply:
I don't like the reasoning in the Airbnb style guide and the ESLint rule, but that's maybe because I'm a pythonista and for us the convention works just fine. They say you should treat everything what's public as public, but I think the reality is [we're all consenting adults](https://mail.python.org/pipermail/tutor/2003-October/025932.html).

> because Airbnb's JavaScript styleguide is considered the holy grail, we don't have to think about it as such

Well, style guides are here to follow and to keep developers aligned despite their different opinions on code. We can express a project-wide opinion, but that, to certain extent, beats the purpose of the style guide.

> because of decaff, we have different settings for some rules anyway

That's an outstanding bug to be resolved 😉 And just FYI, [there are no exceptions](https://github.com/apiaryio/dredd-transactions/blob/master/.eslintrc.js) on this particular repo except one rule you despised and I adored (but I was lazy to fight, even though the style guide was on my side).

-----

That said, I like following the most as well:

```js
// For the purpose of unit tests
compile._compileBody = compileBody;
compile._hasMultipartBody = hasMultipartBody;
```

We can disable the rule on purpose in our projects and if future maintainers disagree with our decision, they can change it. We can link this discussion from a comment in the `.eslintrc.js` file as a documentation of the decision.